src/core/lib/iomgr/network_status_tracker.c

```diff
@@ -40,58 +40,72 @@ typedef struct endpoint_ll_node {
   struct endpoint_ll_node *next;
 } endpoint_ll_node;
 
-static endpoint_ll_node *head = NULL;
-static gpr_mu g_endpoint_mutex;
+typedef struct {
+  gpr_mu mu;
+  endpoint_ll_node *head;
+} shard_type;
+
+#define SHARD_COUNT 32
+static shard_type g_shards[SHARD_COUNT];
 
 void grpc_network_status_shutdown(void) {
-  if (head != NULL) {
-    gpr_log(GPR_ERROR,
-            "Memory leaked as not all network endpoints were shut down");
+  for (size_t i = 0; i < SHARD_COUNT; i++) {
+    shard_type *s = &g_shards[i];
+    if (s->head != NULL) {
+      gpr_log(GPR_ERROR,
+              "Memory leaked as not all network endpoints were shut down");
+    }
   }
-  gpr_mu_destroy(&g_endpoint_mutex);
 }
 
 void grpc_network_status_init(void) {
-  gpr_mu_init(&g_endpoint_mutex);
+  for (size_t i = 0; i < SHARD_COUNT; i++) {
+    shard_type *s = &g_shards[i];
+    gpr_mu_init(&s->mu);
+    s->head = NULL;
+  }
   // TODO(makarandd): Install callback with OS to monitor network status.
 }
 
 void grpc_destroy_network_status_monitor() {
-  for (endpoint_ll_node *curr = head; curr != NULL;) {
-    endpoint_ll_node *next = curr->next;
-    gpr_free(curr);
-    curr = next;
+  for (size_t i = 0; i < SHARD_COUNT; i++) {
+    shard_type *s = &g_shards[i];
+    gpr_mu_destroy(&s->mu);
   }
-  gpr_mu_destroy(&g_endpoint_mutex);
+}
+
+// pointer hash function copied from timer_generic.c
```

Comment:
Can I (beg perhaps) that this be moved to useful.h... it's getting duplicated around a bit too much.

```c
#define GPR_HASH_POINTER(x, range) (((x) >> 4) ^ ((x) >> 9) ^ ((x) >> 14)) % (range)
```